### ISSUE TITLE
hot-reloading - State of the Union

### ISSUE THREAD

**Issue by ORESoftware** [systemjs/systemjs#1546]
I wrote this article and I don't mention anything about SystemJS - not because I don't want to, but I because I have no experience with SystemJS. I would like to incorporate SystemJS into the article - I hear hot-reloading is easier with SystemJS than Webpack:

https://medium.com/@the1mills/hot-reloading-with-react-requirejs-7b2aa6cb06e1

The truth is - hot reloading is so easy to implement with RequireJS - it's very very difficult to implement with Webpack. Dan Abramov is on attempt # 3 with ReactHotLoader3.

Is hot reloading possible with SystemJS? The reason it's so damn easy with RequireJS is because it doesn't "concatenate" your modules in the browser; every module is an isolated JS object.

I will try out SystemJS - but before I do so it would be nice to find out information about how possible hot-reloading is with this module loader.

**Comment by guybedford** [comment]
Perhaps @alexisvincent can help you here.

**Comment by alexisvincent** [comment]
@ORESoftware Checkout [systemjs-hmr](https://github.com/alexisvincent/systemjs-hmr). It's got a pretty good write up of the state of affairs. 

Currently the most popular client facing hmr implementation is [systemjs-hot-reloader](https://github.com/capaj/systemjs-hot-reloader), which as of a few months ago uses [systemjs-hmr](https://github.com/alexisvincent/systemjs-hmr) under the hood. 

I've also rewritten HMR [here](https://github.com/alexisvincent/systemjs-hmr/blob/master/lib/next.js), which will be replacing the current default reloading code. As you can probably see, HMR is dead simple in SystemJS. The file is 239 lines long, and most of that is comments... So feel free to read through it :) 

Note, I'm working on [systemjs-tools](https://github.com/alexisvincent/systemjs-tools), which, once I've added docs, provides a really good dev/HMR experience :)

**Comment by alexisvincent** [comment]
A note on Dan's react HMR... It's still a necessary addition on top of the hmr primitive. What he's achieving there is react state persistence, not hmr itself. So even for SystemJS hmr and the one you wrote for requireJS... You will still need his addition.

**Comment by ORESoftware** [comment]
@alexisvincent thanks, yeah I think the "asynchronous module loaders", namely SystemJS and RequireJS make hot-reloading much simpler to implement. Is there any way you could write a five sentence blurb that explains:

(1) why it's easy to implement hot-reloading with SystemJS
(2) how robust the hot-reloading implementation is
(3) what are the limitations (if any) of the current implementation

if I get those I will put them in the article. I will do my own legwork given what you said and the links provided but I doubt I will ever say it as well as you would!

**Comment by Mosho1** [comment]
I've been using `systemjs-hot-reloader` for over a year in several commercial products' development environment and apart from a few small wrinkles (none recent), it has served me very well. Like @alexisvincent said, `react-hot-loader` is for state persistence across React components and not hot reloading, which is a feature webpack provides (and I think falls short of the SystemJS solution regardless).

**Comment by alexisvincent** [comment]
@ORESoftware 

1. SystemJS exposes a module registry, so you can do things like:
```js
System.registry.get('app/index.js')
System.registry.delete('app/index.js')
System.registry.set('app/index.js@hot-reload-instance', System.newModule({
    someExport: true
}))
```
exposing the registry like this means as a hot-reloader implementor, all I have to do is given a module, find all the other modules that transitively import it via the [trace api](https://github.com/ModuleLoader/es-module-loader/blob/master/README.md#tracing-api), `System.registry.delete` them, and then import the root entry point module. Thats it! State persistence features are a simple addition to that. Where before deleting a module, I check if it imports `@hot`, if so, I save a reference to the old module instance in the registry via a name like `oldModule@hot`, and then if a module imports '@hot', I normalize that to `importersName@hot`. So really, really simple.

2. It's super robust. There's almost no error handling that I have to do, since SystemJS itself is robust. Last night I did some robustness testing and I couldn't break it. I think because it's such a simple system, there's so little that can break, which makes it robust.

3. I can't think of any. There's some features I would still like to implement so checkout the [roadmap](https://github.com/alexisvincent/systemjs-hmr#roadmap) and [issues](https://github.com/alexisvincent/systemjs-hmr/issues).

**Comment by alexisvincent** [comment]
Note, that at the moment there't 2 hot reloading implementations in the wild. 

`systemjs-hmr/next` and `systemjs-hmr`, `systems-hot-reload` uses `systemjs-hmr`. As soon as I've built in backwards compatibility for the old `__unload` and `__reload` state persistence mechanisms into `systemjs-hmr/next`, `systemjs-hot-reloader` will switch to `systemjs-hmr/next`.

At the moment, both expose an interchangeable API. `System.reload(file, options)` so folks can use either.

**Comment by ORESoftware** [comment]
@alexisvincent that's great info - I guess the only thing I have left to ask ftm is - in your owns words how would you explain why it's comparatively easy to hot-reload systemjs modules vs webpack modules?

**Comment by alexisvincent** [comment]
I've never looked properly at webpacks module format so I'm really not the best person to ask here. But at a guess, I would say I think its because SystemJS places an emphasis on loading modules, which have clear boundaries. Thus the water is not messy when removing a module since its really just case of `remove old module`, `insert new one`. From what I've seen of webpack, everything is traditionally executed in the global script space without clear module boundaries. 

In short, SystemJS thinks in `modules`, webpack thinks in `execution context`.

Could be wrong here but thats my understanding of how webpack does module loading.

**Comment by ORESoftware** [comment]
@alexisvincent Ok thanks, so I think hot-reloading SystemJS is easy for the same reason that RequireJS makes it easy:

(1) asynchronous module format from ground up
(2) each module is isolated - modules are not concatenated into a build (in development mode)

Thus the hot-reloading problem just boils down to a JavaScript reference problem - you have to update all existing references to the old module with references to the new module.
